File: dbm-services/sqlserver/db-tools/dbactuator/pkg/util/validate/validate.go
/*
 * TencentBlueKing is pleased to support the open source community by making 蓝鲸智云-DB管理系统(BlueKing-BK-DBM) available.
 * Copyright (C) 2017-2023 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

// Package validate TODO
package validate

import (
	"fmt"
	"reflect"
	"strings"
	"time"

	"dbm-services/sqlserver/db-tools/dbactuator/pkg/util"

	"github.com/go-playground/locales/en"
	ut "github.com/go-playground/universal-translator"
	"github.com/go-playground/validator/v10"
	en_translations "github.com/go-playground/validator/v10/translations/en"
	"github.com/pkg/errors"
)

// ValidateEnums TODO
// make validate tag work with enums tag
// 避免 validate oneof 和 swagger enums 写 2 份重复的校验和文档
// example: Method string `validate:"required,enums" enums:"post,get" json:"method"`
func ValidateEnums(f validator.FieldLevel) bool {
	fieldValue := f.Field().String()
	fieldName := f.StructFieldName()
	// get StructField
	sf, _ := f.Parent().Type().FieldByName(fieldName)
	// get tag value from tag_field enums
	tagValue := sf.Tag.Get(TagEnum)
	enumsValues := strings.Split(tagValue, ",")
	if util.StringsHas(enumsValues, fieldValue) {
		return true
	} else {
		return false
	}
}

// GoValidateStructSimple TODO
// 简单校验 struct，不涉及逻辑
// 如果 struct 上有 tag validate:"enums"，必须启用enum=true校验
func GoValidateStructSimple(v interface{}, enum bool) error {
	validate := validator.New()
	if enum {
		_ = validate.RegisterValidation("enums", ValidateEnums)
	}
	if err := validate.Struct(v); err != nil {
		return err
	}
	return nil
}

// TagEnum TODO
const TagEnum = "enums"

// GoValidateStruct v 不能是Ptr
func GoValidateStruct(v interface{}, enum bool, charset bool) error {
	validate := validator.New()
	uni := ut.New(en.New())
	trans, _ := uni.GetTranslator("en")
	// 提示时显示 json 字段的名字
	validate.RegisterTagNameFunc(func(fld reflect.StructField) string {
		// name := fld.Tag.Get("json")
		name := strings.SplitN(fld.Tag.Get("json"), ",", 2)[0]
		if name == "-" {
			return ""
		}
		return name
	})
	if err := en_translations.RegisterDefaultTranslations(validate, trans); err != nil {
		return err
	}

	if enum {
		_ = validate.RegisterValidation(TagEnum, ValidateEnums)
	}
	if charset {
		_ = validate.RegisterValidation("checkCharset", validCharSet)
	}
	_ = validate.RegisterValidation("time", validateTimeStr)
	if err := validate.Struct(v); err != nil {
		return translateErr2Msg(v, trans, err)
	}
	return nil
}

// translateErr2Msg v 不能是Ptr
func translateErr2Msg(v interface{}, trans ut.Translator, err error) error {
	var errStr []string
	_, ok := err.(*validator.InvalidValidationError)
	if ok {
		return fmt.Errorf("param error:%s", err.Error())
	}
	for _, vErr := range err.(validator.ValidationErrors) {
		if vErr.Tag() == TagEnum {
			errmsg := ""
			// errmsg := customEnumTransFunc(vErr, v)
			if vErr.Param() == "" {
				sf, _ := reflect.TypeOf(v).FieldByName(vErr.StructField())
				tagValue := sf.Tag.Get(TagEnum)
				errmsg = fmt.Sprintf("%s must be one of [%s]", vErr.Field(), tagValue)
			} else {
				errmsg = vErr.Param()
			}
			errStr = append(errStr, errmsg)
			continue
		}
		errStr = append(errStr, vErr.Translate(trans))
	}
	return errors.New(strings.Join(errStr, " || "))
}

func validCharSet(f validator.FieldLevel) bool {
	v := f.Field().String()
	return util.HasElem(v, []string{"default", "utf8mb4", "utf8", "latin1", "gb2312", "gbk", "binary", "gb18030"})
}

// validateTimeStr TODO
// 验证时间字符串 "09:00:00" 这种
func validateTimeStr(f validator.FieldLevel) bool {
	v := f.Field().String()
	if strings.TrimSpace(v) == "" {
		return true
	}
	_, err := time.Parse("15:04:05", v)
	return err == nil
}

File: dbm-ui/backend/flow/utils/sqlserver/validate.py
"""
TencentBlueKing is pleased to support the open source community by making 蓝鲸智云-DB管理系统(BlueKing-BK-DBM) available.
Copyright (C) 2017-2023 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at https://opensource.org/licenses/MIT
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
import ipaddress
from dataclasses import dataclass
from typing import List

from backend.flow.utils.sqlserver.sqlserver_act_payload import SqlserverActPayload
from backend.flow.utils.sqlserver.sqlserver_db_meta import SqlserverDBMeta
from backend.flow.utils.sqlserver.sqlserver_host import Host


@dataclass
class SqlserverInstance:
    """
    sqlserver实例信息基类
    @attributes ip 实例ip，ipv4格式
    @attributes port 实例端口
    @attributes bk_cloud_id 机器所在云区域
    @attributes is_new 是否是这次加入的
    """

    __dataclass_fields__ = None
    host: str
    port: int
    bk_cloud_id: int
    is_new: bool

    def __init__(self, **kwargs):
        for field in SqlserverInstance.__dataclass_fields__:
            setattr(self, field, kwargs.get(field))


@dataclass
class SqlserverCluster:
    """
    sqlserver实例信息基类
    @attributes ip 机器ip，ipv4格式
    @attributes bk_cloud_id 机器所在云区域
    """

    __dataclass_fields__ = None
    port: int
    immutable_domain: str

    def __init__(self, **kwargs):
        for field in SqlserverCluster.__dataclass_fields__:
            setattr(self, field, kwargs.get(field))


@dataclass
class ValidateHandler:
    def __init__(self):
        self.__dataclass_fields__ = None

    def __post_init__(self):
        for key, value in self.__dataclass_fields__.items():
            validate_func = value.metadata.get("validate")
            if validate_func:
                validate_func(getattr(self, key))


def validate_get_payload_func(value: str) -> None:
    """
    判断get_payload_func变量
    """
    if not hasattr(SqlserverActPayload, value):
        raise ValueError(f"There is no {value} method in the SqlserverActPayload class")


def validate_get_dbmeta_func(value: str) -> None:
    """
    判断get_dbmeta_func变量
    """
    if not hasattr(SqlserverDBMeta, value):
        raise ValueError(f"There is no {value} method in the SqlserverDBMeta class")


def validate_hosts(value: List[Host]) -> None:
    """
    判断传入的hosts类型变量
    """
    if not isinstance(value, list):
        raise ValueError("ips variable is not a list")
    for item in value:
        if not isinstance(item, Host):
            raise ValueError(f"/{item}/ One of the elements in the list is in the wrong format")
        if not ipaddress.ip_address(item.ip):
            raise ValueError(f"/{item.ip}/ is not ipv4")
        if not isinstance(item.bk_cloud_id, int) or item.bk_cloud_id < 0:
            raise ValueError(f"/{item.bk_cloud_id}/ is illegal-value")


def validate_instances(value: List[SqlserverInstance]) -> None:
    """
    判断传入的SqlserverInstance类型变量
    """
    if not isinstance(value, list):
        raise ValueError("ips variable is not a list")
    for item in value:
        if not isinstance(item, SqlserverInstance):
            raise ValueError(f"/{item}/ One of the elements in the list is in the wrong format")
        if not ipaddress.ip_address(item.ip):
            raise ValueError(f"/{item.ip}/ is not ipv4")
        if not isinstance(item.port, int) or item.port <= 0 or item.port >= 65535:
            raise ValueError(f"/{item.port}/ is is illegal-port-value")
        if not isinstance(item.bk_cloud_id, int) or item.bk_cloud_id < 0:
            raise ValueError(f"/{item.bk_cloud_id}/ is illegal-bk-cloud_id-value")
